sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,545 @@
+# -*- coding: utf-8 -*-
+"""Ordering Points To Identify the Clustering Structure (OPTICS)
+
+These routines execute the OPTICS algorithm, and implement various
+cluster extraction methods of the ordered list.
+
+Authors: Shane Grigsby <refuge@rocktalus.com>
+         Amy X. Zhang <axz@mit.edu>
+License: BSD 3 clause
+"""
+
+import scipy as sp
+import numpy as np
+from ..utils import check_array
+from ..neighbors import BallTree
+from ..base import BaseEstimator, ClusterMixin
+from ..metrics.pairwise import pairwise_distances
+from ._optics_inner import quick_scan
+
+
+class SetOfObjects(BallTree):
+
+    """Build balltree data structure with processing index from given data
+    in preparation for OPTICS Algorithm
+
+    Parameters
+    ----------
+    data_points: array [n_samples, n_features]"""
+
+    def __init__(self, data_points, metric, **kwargs):
+
+        super(SetOfObjects, self).__init__(data_points,
+                                           metric=metric, **kwargs)
+
+        self._n = len(self.data)
+        self.metric = metric
+        # Start all points as 'unprocessed' ##
+        self._processed = sp.zeros((self._n, 1), dtype=bool)
+        self.reachability_ = sp.ones(self._n) * sp.inf
+        self.core_dists_ = sp.ones(self._n) * sp.nan
+        self._index = sp.array(range(self._n))
+        # Start all points as noise ##
+        self._cluster_id = -sp.ones(self._n, dtype=int)
+        self._is_core = sp.zeros(self._n, dtype=bool)
+        self.ordering_ = []
+
+
+# OPTICS helper functions; these should not be public #
+def _prep_optics(self, min_samples):
+    self.core_dists_[:] = self.query(self.get_arrays()[0],
+                                     k=min_samples)[0][:, -1]
+
+
+def _build_optics(setofobjects, epsilon):
+    # Main OPTICS loop. Not parallelizable. The order that entries are
+    # written to the 'ordering_' list is important!
+    for point in setofobjects._index:
```

Comment:
this will push a bunch of lines to 80 character length... is it important to add the underscores? If so, I could rename to "objects_tree" or similar to get in compliance...